src/size.c

```diff
@@ -0,0 +1,346 @@
+
+  int* sel_data = LOGICAL(sel);
+
+  for (R_len_t i = 0; i < n; ++i, ++data) {
+    int j = *data;
+
+    if (j == NA_INTEGER) {
+      Rf_errorcall(R_NilValue, "Can't subset with a mix of negative indices and missing values");
+    }
+    if (j >= 0) {
+      if (j == 0) {
+        continue;
+      } else {
+        Rf_errorcall(R_NilValue, "Can't subset with a mix of negative and positive indices");
+      }
+    }
+
+    j = -j;
+    if (j > vec_n) {
+      stop_bad_index_length(vec_n, j);
+    }
+
+    sel_data[j - 1] = 0;
+  }
+
+  SEXP out = lgl_as_index(sel, x);
+
+  UNPROTECT(1);
+  return out;
+}
+
+static SEXP int_filter_zero(SEXP index, R_len_t n_zero) {
+  R_len_t n = vec_size(index);
+  const int* data = INTEGER_RO(index);
+
+  SEXP out = PROTECT(Rf_allocVector(INTSXP, n - n_zero));
+  int* out_data = INTEGER(out);
+
+  for (R_len_t i = 0; i < n; ++i, ++data) {
+    int elt = *data;
+    if (elt != 0) {
+      *out_data = elt;
+      ++out_data;
+    }
+  }
+
+  UNPROTECT(1);
+  return out;
+}
+
+static SEXP lgl_as_index(SEXP i, SEXP x) {
+  R_len_t n = Rf_length(i);
+  R_len_t n_vec = vec_size(x);
+
+  if (n == n_vec) {
+    return r_lgl_which(i, true);
+  }
+
+  /* A single `TRUE` or `FALSE` index is recycled to the full vector
+   * size. This means `TRUE` is synonym for missing index (i.e. no
+   * subsetting) and `FALSE` is synonym for empty index.
+   *
+   * We could return the missing argument as sentinel to avoid
+   * materialising the index vector for the `TRUE` case but this would
+   * make `vec_as_index()` an option type just to optimise a rather
+   * uncommon case.
+  */
+  if (n == 1) {
+    int elt = *LOGICAL(i);
+    if (elt == NA_LOGICAL) {
+      SEXP out = PROTECT(Rf_allocVector(INTSXP, n_vec));
+      r_int_fill(out, NA_INTEGER);
+      UNPROTECT(1);
+      return out;
+    } else if (elt) {
+      SEXP out = PROTECT(Rf_allocVector(INTSXP, n_vec));
+      r_int_fill_seq(out, 1);
+      UNPROTECT(1);
+      return out;
+    } else {
+      return vctrs_shared_empty_int;
+    }
+  }
+
+  Rf_errorcall(R_NilValue,
+               "Logical indices must have length 1 or be as long as the indexed vector.\n"
+               "The vector has size %d whereas the index has size %d.",
+               n, vec_size(x));
+}
+
+static SEXP chr_as_index(SEXP i, SEXP x) {
+  SEXP nms = Rf_getAttrib(x, R_NamesSymbol);
+
+  if (nms == R_NilValue) {
+    Rf_errorcall(R_NilValue, "Can't use character to index an unnamed vector.");
+  }
+
+  i = PROTECT(Rf_match(nms, i, NA_INTEGER));
+
+  if (r_int_any_na(i)) {
+    Rf_errorcall(R_NilValue, "Can't index non-existing elements.");
+  }
+
+  UNPROTECT(1);
+  return i;
+}
+
+SEXP vec_as_index(SEXP i, SEXP x) {
+  switch (TYPEOF(i)) {
+  case INTSXP: return int_as_index(i, x);
+  case LGLSXP: return lgl_as_index(i, x);
+  case STRSXP: return chr_as_index(i, x);
+  case SYMSXP: if (i == R_MissingArg) return i;
+
+  // Do we really want to forbid numeric indices here (> 2^31)?
```

Comment:
Not really

Reply:
This was a comment from @krlmlr in the R code. I think we should cast the doubles to ints for sure, but if we want to support long vectors this will make the code more complicated. We'll need to macro-template in many places. The hashtab implementations will need to be duplicated because the long vector indices take twice the memory.

Reply:
Ok, that makes sense. We'll tackle long vectors at some future point.